gettsim/taxes/zve.py

```diff
@@ -318,25 +319,64 @@ def vorsorge2010(tax_unit, tb):
     # calculate x% of relevant employer and employee contributions
     # then subtract employer contributions
     # also subtract health + care + unemployment insurance contributions
-    altersvors2010 = ~tax_unit["child"] * (
-        (0.6 + 0.02 * (np.minimum(tb["yr"], 2025) - 2005)) * (12 * rvbeit_vors)
-        - (12 * 0.5 * rvbeit_vors)
-    )
+    altersvors2010 = ~tax_unit["child"] * vorsorge_year_faktor(tb["yr"]) * (
+        12 * rvbeit_vors
+    ) - (12 * 0.5 * rvbeit_vors)
+
     # These you get anyway ('Basisvorsorge').
     sonstigevors2010 = 12 * (tax_unit["pvbeit"] + 0.96 * tax_unit["gkvbeit"])
     # maybe add avbeit, but do not exceed 1900€.
     sonstigevors2010 = np.maximum(
         sonstigevors2010,
         np.minimum(sonstigevors2010 + 12 * tax_unit["avbeit"], tb["vors_sonst_max"]),
     )
+    return altersvors2010.astype(int) + sonstigevors2010.astype(int)
 
-    vorsorge2010 = altersvors2010 + sonstigevors2010
 
-    return vorsorge2010.astype(int)
+def vorsorge2005(tax_unit, tb):
+    """ Vorsorgeaufwendungen pre 2010
+    Pension contributions are accounted for up to €20k.
+    From this, a certain share can actually be deducted,
+    starting with 60% in 2005.
+    Other deductions are just added, up to a ceiling of 2400 p.a. for standard employees.
 
+    Background: https://bit.ly/32oqCQq
+    """
+    tb["max_altersvors_2005"] = 20000
+    tb["max_sonstige_vors_2005"] = 2400
+    rvbeit_vors_max = np.minimum(
+        tb["max_altersvors_2005"] * vorsorge_year_faktor(tb["yr"]),
+        12 * tax_unit["rvbeit"] * 2,
+    )
+    # intermediate step
+    altersvors2005_int = ~tax_unit["child"] * (
+        vorsorge_year_faktor(tb["yr"]) * (12 * 2 * tax_unit["rvbeit"])
+        - (12 * tax_unit["rvbeit"])
+    ).astype(int)
+    altersvors2005 = np.minimum(rvbeit_vors_max, altersvors2005_int)
```

Comment:
My first draft given the explanations on the website you cited would be:

```python
altersvors2005 =
    (0.6 + 0.02 * (min(tb["yr"], 2025) - 2005)) 
    * np.minimum(tb["max_altersvors_2005"], 12 * tax_unit["rvbeit"] * 2)
    - 12 * tax_unit["rvbeit"]
```

where the factor is only once multiplied with the min(20000,2*rvbeit).

I think it leads to different results when 2*rvbeit (+ possibly private retirement provisions) gets close to the 20000. 

I might be wrong.

Reply:
you are right, the maximum threshold is fix.

Reply:
Even with the added commit 3910c23 it looks wrong to me.

Now you have altersvors2005 = `min(A,B)` with `A = min(20k, 2*RV)` and `B = f*(2*RV)-RV`.

From the linked website it should simply be altersvors2005 = `f*min(20k,2*RV)-RV`.

Reply:
I'm not sure that source is 100% correct. I just found this one: https://www.smartsteuer.de/online/lexikon/g/guenstigerpruefung/ . They even provide the test cases ;)

Reply:
Which one of the examples in your new link is for the 2005 rule? 

The table under
> Abziehbar nach § 10 Abs. 3 EStG

shows the rule I sketched.

Reply:
OK the factor needs definitely to be applied to the threshold as well. the 2005 and 2010 rules actually only differ in their treatment of other contributions